src/hotspot/share/memory/metaspace.cpp

```diff
@@ -589,8 +589,28 @@ bool Metaspace::class_space_is_initialized() {
 // On error, returns an unreserved space.
 ReservedSpace Metaspace::reserve_address_space_for_compressed_classes(size_t size) {
 
+  // First, attempt to reserve the class space such that zero-based addressing is possible.
+  // If the heap has also be allocated in low address regions, this will - with a high probability -
+  // cause the reserved range to nestle alongside the heap.
+  {
+    // First try for zero-base zero-shift (lower 4G); failing that, try for zero-based with max shift (lower 32G)
+    constexpr int num_tries = 8;
```

Comment:
num_tries should be computed instead of hard-coded.

Reply:
Why? In pre-existing code that does similar things, we always hardcode them implicitly (typically by attempt-mapping from A->B in hardcoded stride C). And how would I calculate it?

Reply:
Hmm, I am a bit unsure about the meaning of the `num_tries` parameter. It looks like Windows and Linux will scan the system's memory map and look for the first hole that's larger than `size`. However, if there are a lot of small holes at lower addresses, then doing 8 tries won't find a large enough block, even though such a block may exist below `unscaled_max`.

The default implementation will try to find a free block with fixed steps. In this case, it seems like 

```
num_tries = (unscaled_max + size - 1) / size;
```

would be more appropriate. Otherwise if `size` changes in the future (to be smaller), again you won't be able to find an appropriate block, even if one exists.

So I am not sure if the caller passing in an arbitrary `num_tries` parameter is a good idea. Maybe the API needs to be redesigned.